source/core/slang-string-util.cpp

```diff
@@ -199,72 +199,95 @@ ComPtr<ISlangBlob> StringUtil::createStringBlob(const String& string)
     return (fromChar == toChar || string.indexOf(fromChar) == Index(-1)) ? string : calcCharReplaced(string.getUnownedSlice(), fromChar, toChar);
 }
 
-/* static */void StringUtil::calcLines(const UnownedStringSlice& textIn, List<UnownedStringSlice>& outLines)
+/* static */UnownedStringSlice StringUtil::extractLine(UnownedStringSlice& ioText)
 {
-    char const* begin = textIn.begin();
-    char const* end = textIn.end();
-
-    char const* cursor = begin;
+    char const*const begin = ioText.begin();
+    char const*const end = ioText.end();
 
-    const char* lineStart = cursor;
+    // If we have hit the end then return the 'special' terminator
+    if (begin == nullptr)
+    {
+        return UnownedStringSlice(nullptr, nullptr);
+    }
 
+    char const* cursor = begin;
     while (cursor < end)
     {
         int c = *cursor++;
         switch (c)
         {
             case '\r': case '\n':
             {
-                outLines.add(UnownedStringSlice(lineStart, cursor - 1));
+                // Remember the end of the line
+                const char*const lineEnd = cursor - 1;
 
                 // When we see a line-break character we need
                 // to record the line break, but we also need
                 // to deal with the annoying issue of encodings,
                 // where a multi-byte sequence might encode
                 // the line break.
-
                 if (cursor < end)
                 {
                     int d = *cursor;
                     if ((c ^ d) == ('\r' ^ '\n'))
                         cursor++;
                 }
-                lineStart = cursor;
-                break;
+
+                ioText = UnownedStringSlice(cursor, end);
+                return UnownedStringSlice(begin, lineEnd);
             }
             default:
                 break;
         }
     }
 
-    if (cursor > lineStart)
-    {
-        outLines.add(UnownedStringSlice(lineStart, cursor));
-    }
+    // There is nothing remaining
+    ioText = UnownedStringSlice(nullptr, nullptr);
+
+    // Could be empty, or the remaining line (without line end terminators of)
+    SLANG_ASSERT(begin <= cursor);
+
+    return UnownedStringSlice(begin, cursor);
 }
 
-/* static */bool StringUtil::areLinesEqual(const UnownedStringSlice& a, const UnownedStringSlice& b)
+/* static */void StringUtil::calcLines(const UnownedStringSlice& textIn, List<UnownedStringSlice>& outLines)
 {
-    List<UnownedStringSlice> slicesA;
-    List<UnownedStringSlice> slicesB;
-
-    calcLines(a, slicesA);
-    calcLines(b, slicesB);
+    outLines.clear();
 
-    const auto linesCount = slicesA.getCount();
-    if (linesCount != slicesB.getCount())
+    UnownedStringSlice text(textIn);
+    while (true)
```

Comment:
With my proposed alternative signature, this would become something like:

```c++
UnownedStringSlice text(textIn);
UnownedStringSlice line;
while(extractLine(text, line))
{
    outLines.add(line);
}
```

Which seems pretty clean, all things considered.